pkg/util/ratelimiter/multiratelimiter.go

```diff
@@ -0,0 +1,195 @@
+	defer rl.lock.Unlock()
+	if rl.state == state {
+		return
+	}
+	if rl.state == StateDisabled {
+		rl.cycle = 0
+		rl.tokens = make([]int, len(rl.policy.LimitForPeriod))
+		rl.startTime = nowFunc()
+	}
+	rl.state = state
+}
+
+// SetStateListener sets a state listener for the RateLimiter
+func (rl *MultiRateLimiter) SetStateListener(listener EventListenerFunc) {
+	rl.lock.Lock()
+	defer rl.lock.Unlock()
+	rl.listener = listener
+}
+
+func (rl *MultiRateLimiter) notifyListener(tm time.Time, state State) {
+	if rl.listener != nil {
+		event := Event{
+			Time:  tm,
+			State: stateStrings[state],
+		}
+		go rl.listener(&event)
+	}
+}
+
+// AcquirePermission acquires a permission from the rate limiter.
+// returns true if the request is permitted and false otherwise.
+// when permitted, the caller should wait returned duration before action.
+func (rl *MultiRateLimiter) AcquirePermission(count []int) (bool, time.Duration, error) {
+	rl.lock.Lock()
+	defer rl.lock.Unlock()
+
+	if rl.state == StateDisabled {
+		return true, 0, nil
+	}
+	if len(count) != len(rl.policy.LimitForPeriod) {
+		return false, 0, fmt.Errorf("MultiRateLimiter policy contain %v tokens but get %v tokens", rl.policy.LimitForPeriod, count)
+	}
+
+	now := nowFunc()
+
+	// max tokens could be permitted(including reserved) from the beginning of current cycle
+	maxTokens := make([]int, len(rl.policy.LimitForPeriod))
+	copy(maxTokens, rl.policy.LimitForPeriod)
+	scale := int(rl.policy.TimeoutDuration/rl.policy.LimitRefreshPeriod) + 1
+	for i := range maxTokens {
+		maxTokens[i] *= scale
+	}
+
+	// current cycle index
+	cycle := int(now.Sub(rl.startTime) / rl.policy.LimitRefreshPeriod)
+
+	// rl.tokens is the number of permitted tokens from the beginning of rl.cycle
+	// tokens is the number of tokens have already been permitted from the beginning
+	// of current cycle. note tokens could be less than zero
+	tokens := make([]int, len(rl.tokens))
+	for i, token := range rl.tokens {
+		newToken := token - (cycle-rl.cycle)*rl.policy.LimitForPeriod[i]
+		if newToken < 0 {
+			newToken = 0
+		}
+		tokens[i] = newToken
+	}
+
+	// reject if already reached the permission limitation
+	for i, token := range tokens {
+		if token >= maxTokens[i] {
+			return false, rl.policy.TimeoutDuration, nil
+		}
+	}
+
+	// permit another token
+	for i := range rl.tokens {
+		rl.tokens[i] = tokens[i] + count[i]
+	}
+	rl.cycle = cycle
+
+	// if there are still free tokens in current cycle
+	allFree := true
+	for i, token := range tokens {
+		if token >= rl.policy.LimitForPeriod[i] {
+			allFree = false
+			break
+		}
+	}
+	if allFree {
+		if rl.state != StateNormal {
+			rl.state = StateNormal
+			rl.notifyListener(now, rl.state)
+		}
+		return true, 0, nil
+	}
+
+	// no free tokens in current cycle, we can permit the request, but we need also
+	// return a duration which the caller must wait before proceed.
+	if rl.state != StateLimiting {
+		rl.state = StateLimiting
+		rl.notifyListener(now, rl.state)
+	}
```

Comment:
In the `MultiRateLimiter`, the state transit logic for `StateNormal` and `StateLimiting` should be different from the `RateLimiter`.

For example, suppose the free tokens is [5, 10] and current state is `StateNormal`. According to current logic:

1. the 1st request asks for [10, 10]: state transit to `StateLimiting` and notify the listener.
2. the 2nd request asks for [1,1]: state transit to `StateNormal` and notify the listener.
3. the 3rd request asks for [10,10],: state transit to `StateLimiting` and notify the listener.
4. ...

The state transit toing and froing between `StateNormal` and `StateLimiting`, which is not desired.

Reply:
As we discussed, I will update this part and remove `StateLimiting ` only keep `StateNormal` and `StateDisable` for `MultiRateLimiter`. Since in multi version of rate limiter, since input of `AcquirePermission` is a array slice and can cause frequent state change.